rockcraft/oci.py

```diff
@@ -203,6 +183,57 @@ def set_env(self, env: List[Dict[str, str]]) -> None:
         _config_image(image_path, params)
         emit.message(f"Environment set to {env_list}", intermediate=True)
 
+    def set_control_data(self, metadata: Dict[str, Any]) -> None:
+        """Create and populate the ROCK's control data folder.
+
+        :param metadata: content for the ROCK's metadata YAML file
+        :type metadata: Dict[str, Any]
+        """
+        emit.progress("Setting the ROCK's Control Data")
+        local_control_data_path = Path(tempfile.mkdtemp())
+
+        # the ROCK control data structure starts with the folder ".rock"
+        control_data_rock_folder = local_control_data_path / Path(".rock")
+        control_data_rock_folder.mkdir(parents=True)
+
+        rock_metadata_file = control_data_rock_folder / Path("metadata.yaml")
+        with open(rock_metadata_file, "w", encoding="utf-8") as rock_meta:
+            yaml.dump(metadata, rock_meta)
```

Comment:
We shouldn't need parents in the mkdir as mkdtemp would already create the path up to that point (we should consider cleaning up mkdtemp, I can look into that in a follow up PR)

A string appended to a path, is a Path; we can also use pathlib.Path's open method instead

```suggestion
        # the ROCK control data structure starts with the folder ".rock"
        control_data_rock_folder = local_control_data_path / ".rock"
        control_data_rock_folder.mkdir()

        rock_metadata_file = control_data_rock_folder / "metadata.yaml"
        with rock_metadata_file.open("w", encoding="utf-8") as rock_meta:
            yaml.dump(metadata, rock_meta)
```